src/aead/chacha20_poly1305.rs

```diff
@@ -278,6 +279,89 @@ mod tests {
                     (CRYPTO_POLY1305_STATE_LEN + 255) / 256);
     }
 
+    // This verifies the encryption functionality provided by ChaCha20_ctr32
+    // is successful when either computed on distinct input/output buffers,
+    // or in-place.
```

Comment:
I suggest this tweaking:

```
    // This verifies the encryption functionality provided by ChaCha20_ctr32
    // is successful when either computed on disjoint input/output buffers,
    // or on overlapping input/output buffers. On some branches of the 32-bit
    // x86 and ARM code the in-place operation fails in some situations where
    // the input/output buffers are not exactly overlapping. Such failures are
    // dependent not only on the degree of overlapping but also the length of
    // the data. `open()` works around that by moving the input data to the
    // output location so that the buffers exactly overlap, for those targets.
    // This test exists largely as a canary for detecting if/when that type of
    // problem spreads to other platforms.
```